### ISSUE TITLE
BF16 and MemoryEffient attention not working on AMD MI250

### ISSUE THREAD

**Issue by chauhang** [pytorch/torchtune#597]
On running torchtume on AMD MI250, the fine-tune completes with ~1.58s/it for LoRA distributed recipe. However, getting following warnings for BF16 and MemoryEfficent attention support:

```
BF16 not supported on this hardware. Setting dtype to float32
...

/torchtune/torchtune/modules/attention.py:208: UserWarning: 1Torch was not compiled with memory efficient attention. (Triggered internally at ../aten/src/ATen/native/transformers/hip/sdp_utils.cpp:515.)
  output = nn.functional.scaled_dot_product_attention(
```

Full trace [here](https://gist.github.com/chauhang/7d633b427f5b3a9c35bcd077a4b412ff)

### Steps
1. Git clone and install torch tune
2. Download the model
3. Run lora recipe using: 
`tune --nnodes 1 --nproc_per_node 8 lora_finetune_distributed --config lora_finetune_distributed`

### Environment

```
python -m torch.utils.collect_env

Collecting environment information...
PyTorch version: 2.4.0.dev20240323+rocm6.0
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.0.32830-d62f6a171

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.4.1 20231218 (Red Hat 11.4.1-3)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.34

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk1_zion_755_ga25447393a1d-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI250X / MI250 (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.0.32830
MIOpen runtime version: 3.0.0
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          256
On-line CPU(s) list:             0-255
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7713 64-Core Processor
CPU family:                      25
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
Stepping:                        1
Frequency boost:                 enabled
CPU(s) scaling MHz:              100%
CPU max MHz:                     2000.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        3992.39
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                  AMD-V
L1d cache:                       4 MiB (128 instances)
L1i cache:                       4 MiB (128 instances)
L2 cache:                        64 MiB (128 instances)
L3 cache:                        512 MiB (16 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-63,128-191
NUMA node1 CPU(s):               64-127,192-255


Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] pytorch-triton-rocm==3.0.0+0a22a91d04
[pip3] torch==2.4.0.dev20240323+rocm6.0
[pip3] torchao-nightly==2024.3.20
[pip3] torchtune==0.0.1
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] pytorch-triton-rocm       3.0.0+0a22a91d04          pypi_0    pypi
[conda] torch                     2.4.0.dev20240323+rocm6.0          pypi_0    pypi
[conda] torchao-nightly           2024.3.20                pypi_0    pypi
[conda] torchtune                 0.0.1                    pypi_0    pypi
```

**Comment by drisspg** [comment]
Memory efficient attention is not support for AMD builds and only bf16 + flash should be enabled for these. I would be curious as to the origin of :
`BF16 not supported on this hardware. Setting dtype to float32` and whether this check works correctly on amd

**Comment by rohan-varma** [comment]
The bf16 issue seems to be coming from https://fburl.com/w7tatfyh, which is a check that's highly intertwined w/CUDA / NVIDIA GPUs at the moment. We'll have to figure out how to test for and enable bf16 support on AMD boxes.

As an aside, I think we should actually fail instead of setting back to fp32, what do you think @kartikayk ?

**Comment by kartikayk** [comment]
Yeh this test needs to be updated. I had an offline chat with @chauhang and she mentioned she'll comment back here once she hears back from AMD on this.

@rohan-varma yeh failure makes sense in this case.

**Comment by supernovae** [comment]
The fix for the bf16 should be in nightly (it was checking for cuda libraries to enable this).

I'd love to see memory efficient flash attention. Looks like there are pending PRs for rocm flash attention fix in PyTorch (tagged for 2.3 release) but i'm not sure the memory efficient code is available.  Training on my 7900xtx with torchtune after removing cuda check i get a warning about flash attention and about 1.6it/s but i was able to complete 4 epoch lora training. Trying qlora now.

**Comment by farshadghodsian** [comment]
I can confirm `dtype: bf16` now works when using the nightly build as the problematic checks in torchtune/utils/precision.py have removed. Previously the `torch.version.cuda` and `packaging.version.parse(torch.version.cuda).release >= (11, 0)` checks were returning False as ROCm implementation of PyTorch instead uses `torch.version.hip`. 

My Llama3-8b-Instruct fine-tining dropped from 9-10hrs using FP32 on the Radeon Pro W7900 to 5-6hrs using BF16. Really happy with the speed up.
